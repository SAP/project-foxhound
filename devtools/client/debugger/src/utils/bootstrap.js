/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { bindActionCreators, combineReducers } from "redux";
import ReactDOM from "react-dom";
const { Provider } = require("react-redux");

import ToolboxProvider from "devtools/client/framework/store-provider";
import flags from "devtools/shared/flags";
const {
  registerStoreObserver,
} = require("devtools/client/shared/redux/subscriber");

const { AppConstants } = ChromeUtils.importESModule(
  "resource://gre/modules/AppConstants.sys.mjs"
);

import { SearchDispatcher } from "../workers/search";
import { PrettyPrintDispatcher } from "../workers/pretty-print";

import configureStore from "../actions/utils/create-store";
import reducers from "../reducers";
import * as selectors from "../selectors";
import App from "../components/App";
import { asyncStore, prefs } from "./prefs";
import { persistTabs } from "../utils/tabs";
const { sanitizeBreakpoints } = require("devtools/client/shared/thread-utils");

let gWorkers;

export function bootstrapStore(client, workers, panel, initialState) {
  const debugJsModules = AppConstants.DEBUG_JS_MODULES == "1";
  const createStore = configureStore({
    log: prefs.logging || flags.testing,
    timing: debugJsModules,
    makeThunkArgs: (args, state) => {
      return { ...args, client, ...workers, panel };
    },
  });

  const store = createStore(combineReducers(reducers), initialState);
  registerStoreObserver(store, updatePrefs);

  const actions = bindActionCreators(
    require("../actions").default,
    store.dispatch
  );

  return { store, actions, selectors };
}

export function bootstrapWorkers(panelWorkers) {
  // The panel worker will typically be the source map and parser workers.
  // Both will be managed by the toolbox.
  gWorkers = {
    prettyPrintWorker: new PrettyPrintDispatcher(),
    searchWorker: new SearchDispatcher(),
  };
  return { ...panelWorkers, ...gWorkers };
}

export function teardownWorkers() {
  gWorkers.prettyPrintWorker.stop();
  gWorkers.searchWorker.stop();
}

/**
 * Create and mount the root App component.
 *
 * @param {ReduxStore} store
 * @param {ReduxStore} toolboxStore
 * @param {Object} appComponentAttributes
 * @param {Array} appComponentAttributes.fluentBundles
 * @param {Document} appComponentAttributes.toolboxDoc
 */
export function bootstrapApp(store, toolboxStore, appComponentAttributes = {}) {
  const mount = getMountElement();
  if (!mount) {
    return;
  }

  ReactDOM.render(
    React.createElement(
      Provider,
      { store },
      React.createElement(
        ToolboxProvider,
        { store: toolboxStore },
        React.createElement(App, appComponentAttributes)
      )
    ),
    mount
  );
}

function getMountElement() {
  return document.querySelector("#mount");
}

// This is the opposite of bootstrapApp
export function unmountRoot() {
  ReactDOM.unmountComponentAtNode(getMountElement());
}

function updatePrefs(state, oldState) {
  const hasChanged = selector =>
    selector(oldState) && selector(oldState) !== selector(state);

  if (hasChanged(selectors.getPendingBreakpoints)) {
    asyncStore.pendingBreakpoints = sanitizeBreakpoints(
      selectors.getPendingBreakpoints(state)
    );
  }

  if (
    oldState.eventListenerBreakpoints &&
    oldState.eventListenerBreakpoints !== state.eventListenerBreakpoints
  ) {
    asyncStore.eventListenerBreakpoints = state.eventListenerBreakpoints;
  }

  if (hasChanged(selectors.getTabs)) {
    asyncStore.tabs = persistTabs(selectors.getTabs(state));
  }

  if (hasChanged(selectors.getXHRBreakpoints)) {
    asyncStore.xhrBreakpoints = selectors.getXHRBreakpoints(state);
  }

  if (hasChanged(selectors.getBlackBoxRanges)) {
    asyncStore.blackboxedRanges = selectors.getBlackBoxRanges(state);
  }
}
