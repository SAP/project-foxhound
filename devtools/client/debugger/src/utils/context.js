/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import {
  getThreadContext,
  getSelectedFrame,
  getCurrentThread,
  hasSource,
  hasSourceActor,
  getCurrentlyFetchedTopFrame,
  hasFrame,
} from "../selectors/index";

// Context encapsulates the main parameters of the current redux state, which
// impact most other information tracked by the debugger.
//
// The main use of Context is to control when asynchronous operations are
// allowed to make changes to the program state. Such operations might be
// invalidated as the state changes from the time the operation was originally
// initiated. For example, operations on pause state might still continue even
// after the thread unpauses.
//
// The methods below can be used to compare an old context with the current one
// and see if the operation is now invalid and should be abandoned. Actions can
// also include a 'cx' Context property, which will be checked by the context
// middleware. If the action fails validateContextAction() then it will not be
// dispatched.
//
// Context can additionally be used as a shortcut to access the main properties
// of the pause state.

// A normal Context is invalidated if the target navigates.

// A ThreadContext is invalidated if the target navigates, or if the current
// thread changes, pauses, or resumes.

export class ContextError extends Error {
  constructor(msg) {
    // Use a prefix string to help `PromiseTestUtils.allowMatchingRejectionsGlobally`
    // ignore all these exceptions as this is based on error strings.
    super(`DebuggerContextError: ${msg}`);
  }
}

export function validateNavigateContext(state, cx) {
  const newcx = getThreadContext(state);

  if (newcx.navigateCounter != cx.navigateCounter) {
    throw new ContextError("Page has navigated");
  }
}

export function validateThreadContext(state, cx) {
  const newcx = getThreadContext(state);

  if (cx.thread != newcx.thread) {
    throw new ContextError("Current thread has changed");
  }

  if (cx.pauseCounter != newcx.pauseCounter) {
    throw new ContextError("Current thread has paused or resumed");
  }
}

export function validateContext(state, cx) {
  validateNavigateContext(state, cx);

  if ("thread" in cx) {
    validateThreadContext(state, cx);
  }
}

export function validateSelectedFrame(state, selectedFrame) {
  const newThread = getCurrentThread(state);
  if (selectedFrame.thread != newThread) {
    throw new ContextError("Selected thread has changed");
  }

  const newSelectedFrame = getSelectedFrame(state, newThread);
  // Compare frame's IDs as frame objects are cloned during mapping
  if (selectedFrame.id != newSelectedFrame?.id) {
    throw new ContextError("Selected frame changed");
  }
}

export function validateBreakpoint(state, breakpoint) {
  // XHR breakpoint don't use any location and are always valid
  if (!breakpoint.location) {
    return;
  }

  if (!hasSource(state, breakpoint.location.source.id)) {
    throw new ContextError(
      `Breakpoint's location is obsolete (source '${breakpoint.location.source.id}' no longer exists)`
    );
  }
  if (!hasSource(state, breakpoint.generatedLocation.source.id)) {
    throw new ContextError(
      `Breakpoint's generated location is obsolete (source '${breakpoint.generatedLocation.source.id}' no longer exists)`
    );
  }
}

export function validateSource(state, source) {
  if (!hasSource(state, source.id)) {
    throw new ContextError(
      `Obsolete source (source '${source.id}' no longer exists)`
    );
  }
}

export function validateSourceActor(state, sourceActor) {
  if (!hasSourceActor(state, sourceActor.id)) {
    throw new ContextError(
      `Obsolete source actor (source '${sourceActor.id}' no longer exists)`
    );
  }
}

export function validateThreadFrames(state, thread, frames) {
  const newThread = getCurrentThread(state);
  if (thread != newThread) {
    throw new ContextError("Selected thread has changed");
  }
  const newTopFrame = getCurrentlyFetchedTopFrame(state, newThread);
  if (newTopFrame?.id != frames[0].id) {
    throw new ContextError("Thread moved to another location");
  }
}

export function validateFrame(state, frame) {
  if (!hasFrame(state, frame)) {
    throw new ContextError(
      `Obsolete frame (frame '${frame.id}' no longer exists)`
    );
  }
}

export function isValidThreadContext(state, cx) {
  const newcx = getThreadContext(state);
  return cx.thread == newcx.thread && cx.pauseCounter == newcx.pauseCounter;
}
