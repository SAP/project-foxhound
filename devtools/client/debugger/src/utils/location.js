/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { getSelectedLocation } from "./selected-location";
import { getSource } from "../selectors/index";

/**
 * Note that arguments can be created via `createLocation`.
 * But they can also be created via `createPendingLocation` in reducer/pending-breakpoints.js.
 * Both will have similar line and column attributes.
 */
export function comparePosition(a, b) {
  return a && b && a.line == b.line && a.column == b.column;
}

export function createLocation({
  source,
  sourceActor = null,

  // Line 0 represents no specific line chosen for action
  line = 0,
  column,
}) {
  return {
    source,
    sourceActor,
    sourceActorId: sourceActor?.id,

    // `line` and `column` are 1-based.
    // This data is mostly coming from and driven by
    // JSScript::lineno and JSScript::column
    // https://searchfox.org/mozilla-central/rev/90dce6b0223b4dc17bb10f1125b44f70951585f9/js/src/vm/JSScript.h#1545-1548
    line,
    column,
  };
}

/**
 * Convert location objects created via `createLocation` into
 * the format used by the Source Map Loader/Worker.
 * It only needs sourceId, line and column attributes.
 */
export function debuggerToSourceMapLocation(location) {
  return {
    sourceId: location.source.id,
    // In case of errors loading the source, we might not have a precise location.
    // Defaults to first line and column.
    line: location.line || 1,
    column: location.column || 0,
  };
}

/**
 * Pending location only need these three attributes,
 * and especially doesn't need the large source and sourceActor objects of the regular location objects.
 *
 * @param {Object} location
 */
export function createPendingSelectedLocation(location) {
  return {
    url: location.source.url,

    line: location.line,
    column: location.column,
  };
}

export function sortSelectedLocations(locations, selectedSource) {
  return Array.from(locations).sort((locationA, locationB) => {
    const aSelected = getSelectedLocation(locationA, selectedSource);
    const bSelected = getSelectedLocation(locationB, selectedSource);

    // Order the locations by line number…
    if (aSelected.line < bSelected.line) {
      return -1;
    }

    if (aSelected.line > bSelected.line) {
      return 1;
    }

    // … and if we have the same line, we want to return location with undefined columns
    // first, and then order them by column
    if (aSelected.column == bSelected.column) {
      return 0;
    }

    if (aSelected.column === undefined) {
      return -1;
    }

    if (bSelected.column === undefined) {
      return 1;
    }

    return aSelected.column < bSelected.column ? -1 : 1;
  });
}

/**
 * Source map Loader/Worker and debugger frontend don't use the same objects for locations.
 * Worker uses 'sourceId' attributes whereas the frontend has 'source' attribute.
 */
export function sourceMapToDebuggerLocation(state, location) {
  // From MapScopes modules, we might re-process the exact same location objects
  // for which we would already have computed the source object,
  // and which would lack sourceId attribute.
  if (location.source) {
    return location;
  }

  // SourceMapLoader doesn't known about debugger's source objects
  // so that we have to fetch it from here
  const source = getSource(state, location.sourceId);
  if (!source) {
    throw new Error(`Could not find source-map source ${location.sourceId}`);
  }

  return createLocation({
    ...location,
    source,
  });
}
