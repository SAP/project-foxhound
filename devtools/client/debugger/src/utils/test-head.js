/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

/**
 * Utils for Jest
 * @module utils/test-head
 */

import { combineReducers } from "redux";
import reducers from "../reducers";
import actions from "../actions";
import * as selectors from "../selectors";
import {
  searchWorker,
  prettyPrintWorker,
  parserWorker,
} from "../test/tests-setup";
import configureStore from "../actions/utils/create-store";
import sourceQueue from "../utils/source-queue";
import { setupCreate } from "../client/firefox/create";

// Import the internal module used by the source-map worker
// as node doesn't have Web Worker support and require path mapping
// doesn't work from nodejs worker thread and break mappings to devtools/ folder.
import sourceMapLoader from "devtools/client/shared/source-map-loader/source-map";

/**
 * This file contains older interfaces used by tests that have not been
 * converted to use test-mockup.js
 */

/**
 * @memberof utils/test-head
 * @static
 */
function createStore(client, initialState = {}, sourceMapLoaderMock) {
  const store = configureStore({
    log: false,
    makeThunkArgs: args => {
      return {
        ...args,
        client,
        sourceMapLoader:
          sourceMapLoaderMock !== undefined
            ? sourceMapLoaderMock
            : sourceMapLoader,
        parserWorker,
        prettyPrintWorker,
        searchWorker,
      };
    },
  })(combineReducers(reducers), initialState);
  sourceQueue.clear();
  sourceQueue.initialize({
    newOriginalSources: sources =>
      store.dispatch(actions.newOriginalSources(sources)),
  });

  store.thunkArgs = () => ({
    dispatch: store.dispatch,
    getState: store.getState,
    client,
    sourceMapLoader,
    panel: {},
  });

  // Put the initial context in the store, for convenience to unit tests.
  store.cx = selectors.getThreadContext(store.getState());

  setupCreate({ store });

  return store;
}

/**
 * @memberof utils/test-head
 * @static
 */
function commonLog(msg, data = {}) {
  console.log(`[INFO] ${msg} ${JSON.stringify(data)}`);
}

function makeFrame({ id, sourceId, thread }, opts = {}) {
  return {
    id,
    scope: { bindings: { variables: {}, arguments: [] } },
    location: { sourceId, line: 4 },
    thread: thread || "FakeThread",
    ...opts,
  };
}

function createSourceObject(filename, props = {}) {
  return {
    id: filename,
    url: makeSourceURL(filename),
    isPrettyPrinted: false,
    isExtension: false,
    isOriginal: filename.includes("originalSource"),
  };
}

function createOriginalSourceObject(generated) {
  const rv = {
    ...generated,
    id: `${generated.id}/originalSource`,
  };

  return rv;
}

function makeSourceURL(filename) {
  return `http://localhost:8000/examples/${filename}`;
}

function createMakeSource() {
  const indicies = {};

  return function(name, props = {}) {
    const index = (indicies[name] | 0) + 1;
    indicies[name] = index;

    // Mock a SOURCE Resource, which happens to be the SourceActor's form
    // with resourceType and targetFront additional attributes
    return {
      resourceType: "source",
      // Mock the targetFront to support makeSourceId function
      targetFront: {
        isDestroyed() {
          return false;
        },
        getCachedFront(typeName) {
          return typeName == "thread" ? { actorID: "FakeThread" } : null;
        },
      },
      // Allow to use custom ID's for reducer source objects
      mockedJestID: name,
      actor: `${name}-${index}-actor`,
      url: `http://localhost:8000/examples/${name}`,
      sourceMapBaseURL: props.sourceMapBaseURL || null,
      sourceMapURL: props.sourceMapURL || null,
      introductionType: props.introductionType || null,
      extensionName: null,
    };
  };
}

/**
 * @memberof utils/test-head
 * @static
 */
let creator;
beforeEach(() => {
  creator = createMakeSource();
});
afterEach(() => {
  creator = null;
});
function makeSource(name, props) {
  if (!creator) {
    throw new Error("makeSource() cannot be called outside of a test");
  }

  return creator(name, props);
}

function makeOriginalSource(source) {
  return {
    id: `${source.id}/originalSource`,
    url: `${source.url}-original`,
    sourceActor: {
      thread: "FakeThread",
    },
  };
}

function makeFuncLocation(startLine, endLine) {
  if (!endLine) {
    endLine = startLine + 1;
  }
  return {
    start: {
      line: startLine,
    },
    end: {
      line: endLine,
    },
  };
}

function makeSymbolDeclaration(name, start, end, klass) {
  return {
    id: `${name}:${start}`,
    name,
    location: makeFuncLocation(start, end),
    klass,
  };
}

/**
 * @memberof utils/test-head
 * @static
 */
function waitForState(store, predicate) {
  return new Promise(resolve => {
    let ret = predicate(store.getState());
    if (ret) {
      resolve(ret);
    }

    const unsubscribe = store.subscribe(() => {
      ret = predicate(store.getState());
      if (ret) {
        unsubscribe();
        // NOTE: memoizableAction adds an additional tick for validating context
        setTimeout(() => resolve(ret));
      }
    });
  });
}

function watchForState(store, predicate) {
  let sawState = false;
  const checkState = function() {
    if (!sawState && predicate(store.getState())) {
      sawState = true;
    }
    return sawState;
  };

  let unsubscribe;
  if (!checkState()) {
    unsubscribe = store.subscribe(() => {
      if (checkState()) {
        unsubscribe();
      }
    });
  }

  return function read() {
    if (unsubscribe) {
      unsubscribe();
    }

    return sawState;
  };
}

function getTelemetryEvents(eventName) {
  return window.dbg._telemetry.events[eventName] || [];
}

function waitATick(callback) {
  return new Promise(resolve => {
    setTimeout(() => {
      callback();
      resolve();
    });
  });
}

export {
  actions,
  selectors,
  reducers,
  createStore,
  commonLog,
  getTelemetryEvents,
  makeFrame,
  createSourceObject,
  createOriginalSourceObject,
  createMakeSource,
  makeSourceURL,
  makeSource,
  makeOriginalSource,
  makeSymbolDeclaration,
  waitForState,
  watchForState,
  waitATick,
};
