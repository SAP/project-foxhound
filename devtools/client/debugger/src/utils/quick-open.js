/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { endTruncateStr } from "./utils";
import { getSourceClassnames, getRelativeUrl } from "./source";

export const MODIFIERS = {
  "@": "functions",
  "#": "variables",
  ":": "goto",
  "?": "shortcuts",
};

export function parseQuickOpenQuery(query) {
  const startsWithModifier =
    query[0] === "@" ||
    query[0] === "#" ||
    query[0] === ":" ||
    query[0] === "?";

  if (startsWithModifier) {
    const modifier = query[0];
    return MODIFIERS[modifier];
  }

  const isGotoSource = query.includes(":", 1);

  if (isGotoSource) {
    return "gotoSource";
  }

  return "sources";
}

export function parseLineColumn(query) {
  const [, line, column] = query.split(":");
  const lineNumber = parseInt(line, 10);
  const columnNumber = parseInt(column, 10);
  if (isNaN(lineNumber)) {
    return null;
  }
  if (isNaN(columnNumber)) {
    return { line: lineNumber };
  }
  // columnNumber here is the user input value which is 1-based.
  // Whereas in location objects, line is 1-based, and column is 0-based.
  return {
    line: lineNumber,
    column: columnNumber - 1,
  };
}

export function formatSourceForList(
  source,
  hasTabOpened,
  isBlackBoxed,
  projectDirectoryRoot
) {
  const relativeUrlWithQuery = `${getRelativeUrl(
    source,
    projectDirectoryRoot
  )}${source.displayURL.search || ""}`;
  const subtitle = endTruncateStr(relativeUrlWithQuery, 100);
  const value = relativeUrlWithQuery;
  return {
    value,
    title: source.shortName,
    subtitle,
    icon: hasTabOpened
      ? "tab result-item-icon"
      : `result-item-icon ${getSourceClassnames(source, null, isBlackBoxed)}`,
    id: source.id,
    url: source.url,
    source,
  };
}

export function formatSymbol(symbol) {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location,
  };
}

export function formatShortcutResults() {
  return [
    {
      value: L10N.getStr("symbolSearch.search.functionsPlaceholder.title"),
      title: `@ ${L10N.getStr("symbolSearch.search.functionsPlaceholder")}`,
      id: "@",
    },
    {
      value: L10N.getStr("symbolSearch.search.variablesPlaceholder.title"),
      title: `# ${L10N.getStr("symbolSearch.search.variablesPlaceholder")}`,
      id: "#",
    },
    {
      value: L10N.getStr("gotoLineModal.title"),
      title: `: ${L10N.getStr("gotoLineModal.placeholder")}`,
      id: ":",
    },
  ];
}
