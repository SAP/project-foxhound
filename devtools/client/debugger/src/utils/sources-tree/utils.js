/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { parse } from "../../utils/url";

/**
 * Get the relative path of the url
 * Does not include any query parameters or fragment parts
 *
 * @param string url
 * @returns string path
 */
export function getRelativePath(url) {
  const { pathname } = parse(url);
  if (!pathname) {
    return url;
  }
  const index = pathname.indexOf("/");
  if (index !== -1) {
    const path = pathname.slice(index + 1);
    // If the path is empty this is likely the index file.
    // e.g http://foo.com/
    if (path == "") {
      return "(index)";
    }
    return path;
  }
  return "";
}
