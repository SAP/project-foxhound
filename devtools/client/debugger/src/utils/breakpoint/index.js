/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { getSourceActorsForSource } from "../../selectors/index";
import { sortSelectedLocations } from "../location";
export * from "./breakpointPositions";

// The ID for a Breakpoint is derived from its location in its Source.
export function makeBreakpointId(location) {
  const { source, line, column } = location;
  const columnString = column || "";
  return `${source.id}:${line}:${columnString}`;
}

export function makeBreakpointServerLocationId(breakpointServerLocation) {
  const { sourceUrl, sourceId, line, column } = breakpointServerLocation;
  const sourceUrlOrId = sourceUrl || sourceId;
  const columnString = column || "";

  return `${sourceUrlOrId}:${line}:${columnString}`;
}

/**
 * Create a location object to set a breakpoint on the server.
 *
 * Debugger location objects includes a source and sourceActor attributes
 * whereas the server don't need them and instead only need either
 * the source URL -or- a precise source actor ID.
 */
export function makeBreakpointServerLocation(state, location) {
  const source = location.source;
  if (!source) {
    throw new Error("Missing 'source' attribute on location object");
  }
  const breakpointLocation = {
    line: location.line,
    column: location.column,
  };
  if (source.url) {
    breakpointLocation.sourceUrl = source.url;
  } else {
    breakpointLocation.sourceId = getSourceActorsForSource(
      state,
      source.id
    )[0].id;
  }
  return breakpointLocation;
}

export function createXHRBreakpoint(path, method, overrides = {}) {
  const properties = {
    path,
    method,
    disabled: false,
    loading: false,
    text: L10N.getFormatStr("xhrBreakpoints.item.label", path),
  };

  return { ...properties, ...overrides };
}

export function getSelectedText(breakpoint, selectedSource) {
  return !!selectedSource && !selectedSource.isOriginal
    ? breakpoint.text
    : breakpoint.originalText;
}

export function sortSelectedBreakpoints(breakpoints, selectedSource) {
  return sortSelectedLocations(breakpoints, selectedSource);
}
