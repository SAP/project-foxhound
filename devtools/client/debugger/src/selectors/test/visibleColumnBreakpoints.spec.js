/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { actions, createStore, makeSource } from "../../utils/test-head";
import { createLocation } from "../../utils/location";

import { features } from "../../utils/prefs";

import {
  getColumnBreakpoints,
  getFirstBreakpointPosition,
} from "../visibleColumnBreakpoints";
import {
  makeMockSource,
  makeMockSourceWithContent,
  makeMockBreakpoint,
} from "../../utils/test-mockup";

function pp(line, column) {
  return {
    location: { sourceId: "foo", line, column },
    generatedLocation: { sourceId: "foo", line, column },
  };
}

function defaultSource() {
  return makeMockSource(undefined, "foo");
}

function bp(line, column) {
  return makeMockBreakpoint(defaultSource(), line, column);
}

const source = makeMockSourceWithContent(
  undefined,
  "foo.js",
  undefined,
  `function foo() {
  console.log("hello");
}
console.log('bye');
`
);

describe("visible column breakpoints", () => {
  it("simple", () => {
    const viewport = {
      start: { line: 1, column: 0 },
      end: { line: 10, column: 10 },
    };
    const pausePoints = { 1: [pp(1, 1), pp(1, 5)], 3: [pp(3, 1)] };
    const breakpoints = [bp(1, 1), bp(4, 0), bp(4, 3)];

    const columnBps = getColumnBreakpoints(
      pausePoints,
      breakpoints,
      viewport,
      source,
      source.content
    );
    expect(columnBps).toMatchSnapshot();
  });

  it("ignores single breakpoints", () => {
    const viewport = {
      start: { line: 1, column: 0 },
      end: { line: 10, column: 10 },
    };
    const pausePoints = { 1: [pp(1, 1), pp(1, 3)], 2: [pp(2, 1)] };
    const breakpoints = [bp(1, 1)];
    const columnBps = getColumnBreakpoints(
      pausePoints,
      breakpoints,
      viewport,
      source,
      source.content
    );
    expect(columnBps).toMatchSnapshot();
  });

  it("only shows visible breakpoints", () => {
    const viewport = {
      start: { line: 1, column: 0 },
      end: { line: 10, column: 10 },
    };
    const pausePoints = { 1: [pp(1, 1), pp(1, 3)], 20: [pp(20, 1)] };
    const breakpoints = [bp(1, 1)];

    const columnBps = getColumnBreakpoints(
      pausePoints,
      breakpoints,
      viewport,
      source,
      source.content
    );
    expect(columnBps).toMatchSnapshot();
  });

  if (!features.codemirrorNext) {
    it("doesnt show breakpoints to the right in CM5", () => {
      const viewport = {
        start: { line: 1, column: 0 },
        end: { line: 10, column: 10 },
      };
      const pausePoints = { 1: [pp(1, 1), pp(1, 15)], 20: [pp(20, 1)] };
      const breakpoints = [bp(1, 1), bp(1, 15)];

      const columnBps = getColumnBreakpoints(
        pausePoints,
        breakpoints,
        viewport,
        source,
        source.content
      );
      expect(columnBps).toMatchSnapshot();
    });
  } else {
    it("show all breakpoints including those to the right in CM6", () => {
      const viewport = {
        start: { line: 1, column: 0 },
        end: { line: 10, column: 10 },
      };
      const pausePoints = { 1: [pp(1, 1), pp(1, 15)], 20: [pp(20, 1)] };
      const breakpoints = [bp(1, 1), bp(1, 15)];

      const columnBps = getColumnBreakpoints(
        pausePoints,
        breakpoints,
        viewport,
        source,
        source.content
      );
      expect(columnBps).toMatchSnapshot();
    });
  }
});

describe("getFirstBreakpointPosition", () => {
  it("sorts the positions by column", async () => {
    const store = createStore();
    const { dispatch, getState } = store;

    const fooSource = await dispatch(
      actions.newGeneratedSource(makeSource("foo1"))
    );

    dispatch({
      type: "ADD_BREAKPOINT_POSITIONS",
      positions: [pp(1, 5), pp(1, 3)],
      source: fooSource,
    });

    const position = getFirstBreakpointPosition(
      getState(),
      createLocation({
        line: 1,
        source: fooSource,
      })
    );

    if (!position) {
      throw new Error("There should be a position");
    }
    expect(position.location.column).toEqual(3);
  });
});
