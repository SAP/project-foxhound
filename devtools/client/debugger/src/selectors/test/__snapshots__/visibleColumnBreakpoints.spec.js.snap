// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visible column breakpoints doesnt show breakpoints to the right 1`] = `
Array [
  Object {
    "breakpoint": Object {
      "disabled": false,
      "generatedLocation": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "id": "breakpoint",
      "location": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "options": Object {},
      "originalText": "text",
      "text": "text",
    },
    "location": Object {
      "column": 1,
      "line": 1,
      "sourceId": "foo",
    },
  },
]
`;

exports[`visible column breakpoints ignores single breakpoints 1`] = `
Array [
  Object {
    "breakpoint": Object {
      "disabled": false,
      "generatedLocation": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "id": "breakpoint",
      "location": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "options": Object {},
      "originalText": "text",
      "text": "text",
    },
    "location": Object {
      "column": 1,
      "line": 1,
      "sourceId": "foo",
    },
  },
  Object {
    "breakpoint": undefined,
    "location": Object {
      "column": 3,
      "line": 1,
      "sourceId": "foo",
    },
  },
]
`;

exports[`visible column breakpoints only shows visible breakpoints 1`] = `
Array [
  Object {
    "breakpoint": Object {
      "disabled": false,
      "generatedLocation": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "id": "breakpoint",
      "location": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "options": Object {},
      "originalText": "text",
      "text": "text",
    },
    "location": Object {
      "column": 1,
      "line": 1,
      "sourceId": "foo",
    },
  },
  Object {
    "breakpoint": undefined,
    "location": Object {
      "column": 3,
      "line": 1,
      "sourceId": "foo",
    },
  },
]
`;

exports[`visible column breakpoints simple 1`] = `
Array [
  Object {
    "breakpoint": Object {
      "disabled": false,
      "generatedLocation": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "id": "breakpoint",
      "location": Object {
        "column": 1,
        "line": 1,
        "source": Object {
          "displayURL": Object {
            "fileExtension": "",
            "filename": "url",
            "group": "",
            "path": "url",
            "search": "",
          },
          "extensionName": null,
          "id": "foo",
          "isExtension": false,
          "isOriginal": false,
          "isPrettyPrinted": false,
          "isWasm": false,
          "longName": "url",
          "shortName": "url",
          "thread": "FakeThread",
          "url": "url",
        },
      },
      "options": Object {},
      "originalText": "text",
      "text": "text",
    },
    "location": Object {
      "column": 1,
      "line": 1,
      "sourceId": "foo",
    },
  },
  Object {
    "breakpoint": undefined,
    "location": Object {
      "column": 5,
      "line": 1,
      "sourceId": "foo",
    },
  },
]
`;
